### DISCUSSION TITLE
Unexpected Common Lisp package behavior

### DISCUSSION THREAD

**Discussion by pamoroso** [Interlisp/medley#1340] [Q&A]
To adapt my Interlisp development workflow to Common Lisp I'm putting together a test case of a simple Common Lisp program, which just calls `IL:CREATEW` from a function `FUN1`, coded across a number of Medley sessions.

However, my understanding of how Medley manages packages is still incomplete and the program doesn't work as I expect. Here are the steps I go through.

In a XCL Exec of a Common Lisp Interlisp Online session I define the file environment of the program and create the coms file:

```
3> (XCL:DEFINE-FILE-ENVIRONMENT PKGTEST :PACKAGE (DEFPACKAGE "PKGTEST" (:USE "LISP" "XCL")) :READTABLE "XCL")
PKGTEST
4> (IL:FILES?)
the File info: PKGTEST
the advice: IL:|NC.LogoutAdvice|
...to be dumped.  want to say where the above go ? Yes
(File info)
PKGTEST  File name: PKGTEST
create new file PKGTEST ? Yes
(advice)
IL:|NC.LogoutAdvice|  Nowhere
6> (IL:MAKEFILE 'IL:PKGTEST)
IL:|{DSK}<home>medley>il>PKGTEST.;1|
```

I can now edit the coms to add the export to the package definition:

```
7> (IL:DC 'IL:PKGTEST)
Editing FILES definition of IL:PKGTEST.
IL:PKGTEST
```

The modified coms looks like this:

![PKGTEST1](https://github.com/Interlisp/medley/assets/10342479/ee96bf1b-c935-427e-8d85-cd1e95bf24af)

and I can save it:

```
11> (IL:MAKEFILE 'IL:PKGTEST)
IL:|{DSK}<home>medley>il>PKGTEST.;2|
```

Next, in a XCL Exec of a fresh session I load the new file and edit the coms to import `IL:CREATEW` :

```
3> (LOAD 'PKGTEST)

; Loading {DSK}<home>medley>il>PKGTEST.;2
; File created  7-Oct-2023 09:12:46
; IL:PKGTESTCOMS
IL:|{DSK}<home>medley>il>PKGTEST.;2|
4> (IL:DC 'IL:PKGTEST)
Editing FILES definition of IL:PKGTEST.
IL:PKGTEST
```

At this point the coms looks like this:

![PKGTEST2](https://github.com/Interlisp/medley/assets/10342479/29279dd3-0bbb-4487-988a-08f5c66a25ab)

Next, I evaluate the package definition to update the running image:

```
5> (DEFPACKAGE "PKGTEST" (:USE "LISP" "XCL") (:EXPORT PKGTEST::FUN1) (:IMPORT IL:CREATEW))
"PKGTEST"
```

Finally, I define the function `PKGTEST:FUN1` that calls `IL:CREATEW`:

```
6> (ED 'PKGTEST:FUN1 '(FUNCTIONS :DONTWAIT))
PKGTEST:FUN1 has no FUNCTIONS definition.
Select a definer to use for a dummy definition.
PKGTEST:FUN1
```

This is the code of `PKGTEST.FUN1`:

![PKGTEST3](https://github.com/Interlisp/medley/assets/10342479/9264c3f1-5776-4ccf-92cc-aa2ddcf5a8cc)

Calling `PKGTEST.FUN1` opens a new window as expected and so I can save the updated file:

```
7> (PKGTEST:FUN1)
#<IL:WINDOW @ 123,12000>
9> (IL:MAKEFILE 'IL:PKGTEST)
IL:|{DSK}<home>medley>il>PKGTEST.;3|
```

Again at a XCL Exec of yet another session I load the file, call `PKGTEST:FUN1`, and get this error:

```
NIL
3> (LOAD 'PKGTEST)

; Loading {DSK}<home>medley>il>PKGTEST.;3
; File created  7-Oct-2023 09:22:33
; IL:PKGTESTCOMS
IL:|{DSK}<home>medley>il>PKGTEST.;3|
4> (PKGTEST:FUN1)
Undefined car of form
PKGTEST:FUN1

5> (APROPOS "FUN1" :PKGTEST)
PKGTEST:FUN1
```

`FUN1` is apparently exported from `PKGTEST`, so why can't I call it via the properly qualified name?

Here is the `FILEPKG` source file:
[PKGTEST.TXT](https://github.com/Interlisp/medley/files/12838367/PKGTEST.TXT)

**Comment by pamoroso** [comment]
A first thought is `XCL:DEFINE-FILE-ENVIRONMENT` doesn't evaluate any of the arguments, so I shouldn't pass `(DEFPACKAGE "PKGTEST" (:USE "LISP" "XCL"))` but `"PKGTEST"`.

**Comment by Anzus** [comment]
I don’t see FUN1 defined in the filecoms for the file, so it isn’t being saved to the file.

**Comment by pamoroso** [comment]
Thanks @Anzus. I loaded the source in a new session and defined the function `FUN1`. Calling `FUN1` still yelds an error as the symbol isn't exported from `PKGTEST`, but this is likely due to the [incorrect definition of the package](https://github.com/orgs/Interlisp/discussions/1340#discussioncomment-7219490):

```
3> (LOAD 'PKGTEST)

; Loading {DSK}<home>medley>il>PKGTEST.;3
; File created  7-Oct-2023 09:22:33
; IL:PKGTESTCOMS
IL:|{DSK}<home>medley>il>PKGTEST.;3|
4> (ED 'PKGTEST:FUN1 '(FUNCTIONS :DONTWAIT))
PKGTEST:FUN1 has no FUNCTIONS definition.
Select a definer to use for a dummy definition.
PKGTEST:FUN1
5> (PKGTEST:FUN1)
Undefined car of form
PKGTEST::CREATEW

6> (APROPOS "CREATEW" :PKGTEST)
PKGTEST::CREATEW
```

**Comment by pamoroso** [comment]
I went through the same steps again, with a couple of differences.

First, I defined the `PKGTEST` package before defining the file environment because, when I later try to save the coms, I get a package not found error. Also, I passed `IL:PKGTEST` as the first argument of `XCL:DEFINE-FILE-ENVIRONMENT` instead of `PKGTEST` as otherwise the symbol is interned in the `XCL-USER` package.

Here's the transcript of the first session

```
3> (DEFPACKAGE "PKGTEST" (:USE "LISP" "XCL"))
"PKGTEST"
4> (XCL:DEFINE-FILE-ENVIRONMENT IL:PKGTEST :PACKAGE "PKGTEST" :READTABLE "XCL")
IL:PKGTEST
5> (IL:FILES?)
the File info: IL:PKGTEST
the advice: IL:|NC.LogoutAdvice|
...to be dumped.  want to say where the above go ? Yes
(File info)
IL:PKGTEST  File name: PKGTEST
create new file PKGTEST ? Yes
(advice)
IL:|NC.LogoutAdvice|  Nowhere
6> (IL:MAKEFILE 'IL:PKGTEST)
IL:|{DSK}<home>medley>il>PKGTEST.;1|
7> (IL:DC 'IL:PKGTEST)
Editing FILES definition of IL:PKGTEST.
IL:PKGTEST
8> (IL:MAKEFILE 'IL:PKGTEST)
IL:|{DSK}<home>medley>il>PKGTEST.;2|
```

When I load the symbolic file `PKGTEST` in a new session I get the error `Can't find package for DEFINE-FILE-INFO reader environment PKGTEST`:

![PKGTEST](https://github.com/Interlisp/medley/assets/10342479/29ca69ae-ce7d-49b7-bc6f-2479848dbe0e)

Here's the symbolic file `PKGTEST`:

[PKGTEST.TXT](https://github.com/Interlisp/medley/files/12841152/PKGTEST.TXT)

If I pass `PKGTEST` instead of `IL:PKGTEST` as the first argument of `XCL:DEFINE-FILE-ENVIRONMENT` I still get the same error `Can't find package for DEFINE-FILE-INFO reader environment PKGTEST` later in the process.

**Comment by Anzus** [comment]
What I’m trying to say is that looking at the files you’ve sent, the COMS variable for PKGTEST doesn’t contain a FUNCTIONS reference for FUN1 so it’s not getting saved to the file. When you load your file into a fresh sys out, FUN1 isn’t there in the new image. After defining FUN1 run (FILES?) again and it’ll ask you if you want to put it somewhere, add it to PKGTEST and then make the file again.

**Comment by pamoroso** [comment]
You're right, I hadn't tried saving `FUN` yet. So I went again through exactly the same steps of the [initial procedure](https://github.com/orgs/Interlisp/discussions/1340#discussion-5711241) except that after step 7 I called `FILES?` and `MAKEFILE` to save `FUN1`. However, when in a fresh session I load the source file and call `(PKGTEST:FUN1)` I get the error:

```
Undefined car of form
PKGTEST::CREATEW
```

Here's the `PKGTEST` source:

[PKGTEST.TXT](https://github.com/Interlisp/medley/files/12845651/PKGTEST.TXT)

Since I imported the `CREATEW` symbol into the `PKGTEST` package I would assume it could be used unqualified in a file set to the `PKGTEST` package. But there may be some issue with the package definition or the makefile environment as [I hinted earlier](https://github.com/orgs/Interlisp/discussions/1340#discussioncomment-7223688).

**Reply by nbriggs** [reply]
You're in package PKGTEST when you're reading, so importing an unqualified CREATEW is unlikely to import the IL:CREATEW that you need.  I would definitely NOT do it this way - I would call IL:CREATEW in the function definition and not import Interlisp symbols into the package.

**Reply by pamoroso** [reply]
Thanks for the insight. I assumed that, once `CREATEW` is imported in `PKGTEST`, reading an unqualified `CREATEW` while in package `PKGTEST` would work, as I thought the symbol would be accessible from the package.

By the way, the [Lyric Common Lisp implementation notes](http://www.bitsavers.org/pdf/xerox/interlisp-d/198706_Lyric/3102464_Lyric_Common_Lisp_Implementation_Notes_Jun87.pdf), on page 38 of the PDF, discuss an example of importing the `CREATEW` and `WINDOWPROP` Interlisp symbols into a Common Lisp package.

**Reply by nbriggs** [reply]
Yes, re page 38 - and notice that it imports IL:CREATEW, not CREATEW.

**Reply by pamoroso** [reply]
In the steps I described here I think I was careful to qualify `IL:CREATEW` when importing it into the `PKGTEST` package. I'd like to figure what I'm doing wrong as I'll likely want to import symbols from other packages in my own code.

**Reply by nbriggs** [reply]
Look carefully at the text of the PKGTEST file - it's unclear to me which DEFPACKAGE is actually going to control the creation of the package - there are multiple definitions.

**Reply by pamoroso** [reply]
It's unclear to me too, I only add the package definitions I mentioned in the steps above.

**Comment by pamoroso** [comment]
@MattHeffron This is the Common Lisp package issue mentioned in the latest weekly meeting, I'd appreciate your feedback.

**Comment by Anzus** [comment]
Also, 
(DEFPACKAGE "PKGTEST" (:USE "LISP" "XCL")
       (:EXPORT FUN1)
       (:IMPORT CREATEW))

Doesn’t say which package to import CREATEW from. What you probably want is
(DEFPACKAGE "PKGTEST" (:USE "LISP" "XCL")
       (:EXPORT FUN1)
       (:IMPORT IL:CREATEW))

**Reply by pamoroso** [reply]
I actually do add the package qualifier when importing `CREATEW` but, since the package is set to Interlisp when editing the coms, SEdit immediately hides the qualifier `IL` as soon as I type a space after the last character of `CREATEW`. Maybe the qualifier gets lost when saving the coms.

**Comment by MattHeffron** [comment]
@pamoroso I was digging around in the source code to see if I could find an easy way to make it work close to the way that you tried. I didn't really see it. What I did, way back in the mid 1980's, was to setup the DEFINE-FILE-INFO with the basic information in the DEFPACKAGE, such as NICKNAMES, PREFIX-NAME, and USE (and EXTERNAL-ONLY for some of my packages). Any other modification of the package was done as part of the COMS in a P command. 
DEFPACKAGE is not defined in the same way as DEFUN, DEFMACRO, DEFPARAMETER, etc. The Interlisp File Package does not detect changes to the definition of a package, so whatever was used in the DEFINE-FILE-INFO will be what's written at the next MAKEFILE.
When the DEFINE-FILE-INFO specified a package other than IL, then the COMS are written in _that_ package. **HOWEVER**, once symbols from IL have been imported into this package, then they will **not** be prefixed in the COMS the next time the file is written. So, when I wanted to IMPORT a symbol into my MP package from the IL package, I used a construct like:
```common-lisp
(IL:RPAQQ IL:MPLOADCOMS
          ((IL:P (IL:|for| IL:S IL:|in|
                        '("\\GETWINDOWHEIGHT" "BITBLT" "BITMAPBIT" "BITMAPCREATE" "BITMAPHEIGHT" 
                                "BITMAPWIDTH" "BKSYSCHARCODE" "BLTSHADE" "CHARCODE" "CHARWIDTH" 
                                "CLEARBUF" "CONCAT" "DEL.PROCESS" "DSPCLIPPINGREGION" "DSPFONT" 
                                "DSPRESET" "FIND-READTABLE" "FIXPXLS" "FONTHEIGHT" "FONTPROP" 
                                "GETMOUSESTATE" "GETSTREAMPROP" "INPUT" "INSIDEP" "INTERRUPTCHAR" 
                                "LASTMOUSEBUTTONS" "LASTMOUSEX" "LASTMOUSEY" "SHRINKW" "TAB" 
                                "THIS.PROCESS" "WHITESHADE" "WINDOWPROP")
                        IL:|do|
                        (IMPORT (FIND-SYMBOL IL:S "IL")
                               "MP"))
                 
                 (IL:* IL:|;;| 
                 "Make sure the symbols already exist before they get loaded from in a KB. These are initially defined in other files.")
                 '(YES NO UNKNOWN PREMISE ANTECEDENT ACTION CURULE QUESTION.ANSWER.RECORD 
                       ANTECEDENT-FOR-PARMS HALFTRACED TRACING DEFAULT ANTECEDENT-IN RULERECORD ASKED
                       TYPE-IN C.UPDATED-BY.FILE PROMPT PERMANENT.CONSTRAINTS TEXT WORKING.GOALS 
                       ACTIVATED REPLAY.ANSWERS ANSWER.OR.CONCLUSION.CHANGED CHANGE.ANSWER POSNUMB 
                       HOURS MINUTES EXPECT LEGALVALS YESNO YESNO.UNKNOWN FROMLEGALVALS RANGE 
                       MULTIVALUED FACT.TABLE DOMAIN.VARIABLES SCRATCH.VARIABLES FACT.TABLES 
                       ROOT-CONTEXT VALUE.TRANSLATIONS DOM.SCR.VARIABLES GOALS PARMS RULEGROUPS 
                       NOTCOMPILEDRULES PRUNEDRULES UPDATED-BY USED-BY *PLAN-FILENAMES* *CPB.BUILT* 
                       RIGHT MIDDLE LEFT ACTIVATED.BY RULESFILENAME *TOP-LEVEL-NAME* TRANS)

                 (IL:* IL:|;;| "Can't just call WINDOWPROP here or else the symbol gets created in the MP package when the COMS are read in, and then when the IMPORT above is done, you get a name conflict.")
                 (FUNCALL (FIND-SYMBOL "WINDOWPROP" "IL")
                        (IL:WFROMDS T T)
                        'IL:PAGE.WAIT-SECONDS 0))
``` 
_etc._
Also, note that the (IMPORT ...) is the **first** thing in the COMS, so that it's done **_before_** the definitions that follow are printed to the file, because they will be printed without the IL: prefix, since when the file is next written, the IMPORT will have been done.
Once I figured this out (back in the _ancient past_) it has worked very well. I hope this helps.

**Reply by pamoroso** [reply]
@MattHeffron Thanks for sharing your experience and insight, I'll study it and experiment further.

In the meantime I'd like to clarify the intent behing the convoluted steps I described here. Since Lisp software is usually developed incrmentally in an exploratory process, I want to devise a workflow for evolving package definitions (e.g. adding and exporting symbols, importing useful symbols) along with the code in such a way that the File Manager keeps track of the changes.

**Reply by pamoroso** [reply]
@MattHeffron I have a few questions and comments on a couple of things you said.

> What I did, way back in the mid 1980's, was to setup the DEFINE-FILE-INFO with the basic information in the DEFPACKAGE [...]

How did you set up the `DEFINE-FILE-INFO`? Via the `IL:MAKEFILE-ENVIRONMENT` property of the filename symbol? `XCL:DEFINE-FILE-ENVIRONMENT`?

> DEFPACKAGE is not defined in the same way as DEFUN, DEFMACRO, DEFPARAMETER, etc. The Interlisp File Package does not detect changes to the definition of a package [...]

Reading the Medley documentation gave me the impression the File Manager notices package changes. For example, [issue 5 of the HOTLINE!](http://www.bitsavers.org/pdf/xerox/interlisp-d/hotline/05_19871016.pdf) newsletter says on page 8 of the PDF:

> XCL provides the function DEFPACKAGE to enable easy interaction with the File Manager.

On page 9 this and the text that follows:

> DEFPACKAGE has been provided to enable easy interaction with the File Manager. DEFPACKAGE can be used in a file's IL:MAKEFILE-ENVIRONMENT property to define the package in which the file is to be read and written.

**Reply by pamoroso** [reply]
@MattHeffron [wrote](https://github.com/orgs/Interlisp/discussions/1340#discussioncomment-7247407):

> What I did, way back in the mid 1980's, was to setup the DEFINE-FILE-INFO with the basic information in the DEFPACKAGE [...]

How did you set up the `DEFINE-FILE-INFO`? With `XCL:DEFINE-FILE-ENVIRONMENT`, `IL:MAKEFILE-ENVIRONMENT`, or some other way?

**Reply by MattHeffron** [reply]
I setup the `DEFINE-FILE-INFO` as the `IL:MAKEFILE-ENVIRONMENT` property on the filename symbol. 
(When I first started needing packages, I think that the `XCL:DEFINE-FILE-ENVIRONMENT` function wasn't there yet. In any case, that was **_before_** the HOTLINE! article you referenced.)

**Comment by rmkaplan** [comment]
Does it make sense to provide a definition for PACKAGES as a FILEPKG command and as a FILEPKG type, like RECORDS, VARS etc.?

**Comment by nbriggs** [comment]
Have a read through https://bitsavers.org/pdf/xerox/interlisp-d/hotline/05_19871016.pdf on packages.  This touches on the issues that @MattHeffron noted above.

**Comment by masinter** [comment]
packages are probably not something you should define incrementally because many parts of the environment assume you can random access a source file knowing only the DEFINE-FILE-INFO at the top of the file. There is no per-funcdtion/per-definer variation supported. You can write such programs and import them if necessary.

I think for the typical incremental development, you don't start with enough of an understanding of the program to predict.  Why import IL:CREATEW -- maybe you save 3 keystrokes in SEdit -- but it adds to the complexity needed for anyone to understand.

(There are a couple of other "gotcha"'s in Common LIsp that are difficult to deal with but not many; the one that comes to mind is #+ and #-.

These things were issues in X3J13 ...

**Comment by Anzus** [comment]
I will echo the comment about not doing package stuff incrementally. I’m afraid that package manipulation was easier with non-resident environments. One other “hack” to handle this is to have a non-Medley-managed file (i.e., a lisp source file that has been edited with a text editor) that has all the package definition/manipulation stuff in pure CL that is loaded before the collection of Medley files, so that all the symbols you need to cross package boundaries are defined before they get created in the wrong place. Looking at things like PKG.LISP in the PCL sources, I suspect this is not that unusual a technique in the pure-CL world, and it’s probably where I picked up the habit. Matt’s solution is more true to the Interlisp way of life.

**Reply by MattHeffron** [reply]
My solution is as Arun noted. The package definition is all in the xxLOAD file for the project. If I need to modify a package definition, I can edit the xxLOADCOMS, write a new xxLOAD file, and then exit, spin up a clean sysout, and re-LOAD the project; and I'm back "in business". (It's so much faster now, vs. then, that it's hardly worth thinking about finding a "more efficient" way to do it.)
Matt

(Hi Arun. I'm not sure if we've communicated since AAAI-91 in Anaheim, when I brought you to visit our "AI Lab" at Beckman!)

**Reply by MattHeffron** [reply]
Another thing ... is that within a "project", whenever I create a new file I just copy the `IL:MAKEFILE-ENVIRONMENT` from an existing file in the project. (Rarely, I had to modify that; but, for me, it was just for cases involving the files mimicking functions in the Gold Hills Common Lisp for the PC.)

**Comment by rmkaplan** [comment]
I agree with Arun, that the way that packages are incorporated into Medley’s residential programming model with managed files is not well thought out, cumbersome, and error prone (also see previous comments in #474).

The intuitive scenario in the simplest case would be that a file FOO that wants to be loaded with unqualified symbols in a package FOOPKG defines that package in its commands FOOCOMS, so that everything is self-contained.  FOO’s DEFINE-FILE-INFO would simply say that FOO uses package FOOPKG, no need for a little file off to the side that exists just to provide the package that the DEFINE-FILE-INFO of FOO can load before FOO itself is read.  

The (or at least one) problem is what in linguistics is called an ordering paradox.  If FOOPKG is specified inside FOOCOMS and appears later in the file, DEFINE-FILE-INFO can’t get to that specification without crossing over lots of symbols that can’t be interpreted without already having obtained that specification.

A possible solution is to arrange it so that DEFINE-FILE-INFO can get to and execute the package specification locally in FOO without first tripping over other symbols, storing other data, evaluating other expressions etc.

So here’s a proposal, based on the way that the EXPORTS filepkg command and GATHER-EXPORTS work for extracting declarations from a collection of source files without loading anything else from those files.

1.  Introduce a filepkg command PACKAGES that puts out code that, when executed, defines all of the properties that currently exist in memory for whatever packages are specified.  E.g. Putting (PACKAGES “FOOPKG") in FOOCOMS would produce code that recreates FOOKPKG on loading.

But: in front of all the FOOKPGG-defining code, also put out a magical comment string “DEFINITION FOR FOOPKG” (or maybe there is just some recognizable part of the code (like (DEFPACKAGE FOOPKG…) that will serve the intended purposes).  That purpose is to be able to find the FOOPKG specification in the file by a simple string search that doesn’t need to interpret any other characters that come before or after it in the file.

2.  Extend DEFINE-FILE-INFO (actually, the function \DO-DEFINE-FILE-INFO) to deal with the case that the file’s package (identified simply by its name, no complicated or redundant expression) doesn’t already exist.  In that case, if it doesn’t see a form to EVAL, it would scan the file for the magical string, execute the defining code to create the package on the fly, and then loads the rest of the file.

In other words, with this setup the DEFINE-FILE-INFO can find and create the needed package before it looks at any other symbols in the file—the package will always be created first.

3.  It might also help to define a FILEPKGTYPE for packages, so that all the properties for a package can be brought up for editing in the obvious way.

A little more complicated is the case of a multi-file application, like TEDIT, where the package is defined in one of the files but is required before any of the others can be loaded.  Each of them could specify its DEFINE-FILE-INFO with a form that indicates the package name and the file it is contained in, and DEFINE-FILE-INFO would do the scan on that other file, again without having to deal with anything else on that file.  This is like the current mode of having a separate little package-file, except that the package does not have to be isolated in its own stand-alone file.

BTW, the HOTLINE also has examples of creating the makefile environment on the property list of the file, and preserving that property explicitly on the file.  I think that the various reader environment elements can be included in the OPTIONS to the original MAKEFILE, and that afterwards they are automatically carried along to new versions.  The first argument of CLEANUP, if it's a list, is also passed down as options to the makefile.  That seems like a more coherent way of interacting with this stuff.
